// react-native-autocomplete-dropdown/lib/commonjs/HOC/withFadeAnimation.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.withFadeAnimation = void 0;
var _react = _interopRequireWildcard(require("react"));
var _reactNative = require("react-native");
var _jsxRuntime = require("react/jsx-runtime");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
const withFadeAnimation = (WrappedComponent, {
  containerStyle
} = {}) => {
  return props => {
    const opacityAnimationValue = (0, _react.useRef)(new _reactNative.Animated.Value(0)).current;
    (0, _react.useEffect)(() => {
      _reactNative.Animated.timing(opacityAnimationValue, {
        duration: 800,
        toValue: 1,
        useNativeDriver: true,
        easing: _reactNative.Easing.bezier(0.3, 0.58, 0.25, 0.99)
      }).start();
    }, [opacityAnimationValue]);
    return /*#__PURE__*/(0, _jsxRuntime.jsx)(_reactNative.Animated.View, {
      style: [containerStyle, {
        opacity: opacityAnimationValue
      }],
      children: /*#__PURE__*/(0, _jsxRuntime.jsx)(WrappedComponent, {
        ...props
      })
    });
  };
};
exports.withFadeAnimation = withFadeAnimation;
